sdk/src/main/java/com/akkaserverless/javasdk/view/ViewBase.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2021 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.akkaserverless.javasdk.view;
+
+import com.akkaserverless.javasdk.impl.view.ViewUpdateEffectImpl;
+
+import java.util.Optional;
+
+/*
+ * FIXME rename this to View when the old annotation is removed
+ */
+/** @param <S> The type of the state for this view. */
+public abstract class ViewBase<S> {
+
+  private Optional<UpdateHandlerContext> updateHandlerContext = Optional.empty();
+
+  /**
+   * Additional context and metadata for an update handler.
+   *
+   * <p>It will throw an exception if accessed from constructor.
+   */
+  protected final UpdateHandlerContext updateHandlerContext() {
+    return updateHandlerContext.orElseThrow(
+        () ->
+            new IllegalStateException(
+                "UpdateHandlerContext is only available when handling an update."));
+  }
+
+  /** INTERNAL API */
+  public final void setUpdateHandlerContext(Optional<UpdateHandlerContext> context) {
+    updateHandlerContext = context;
+  }
+
+  protected final UpdateEffect.Builder<S> updateEffects() {
+    return ViewUpdateEffectImpl.builder();
+  }
+
+  /**
+   * Construct the effect that is returned by the command handler. The effect describes next
+   * processing actions, such as emitting events and sending a reply.
+   *
+   * @param <S> The type of the state for this entity.
+   */
+  public interface UpdateEffect<S> {
+
+    interface Builder<S> {
+
+      UpdateEffect<S> updateState(S newState);
+
+      /** Ignore this event (and continue to process the next). */
+      UpdateEffect<S> ignore();
+
+      /**
+       * Trigger an error for the event. FIXME what happens on error, retry of the same event?
```

Comment:
yes, offset is not stored, projection is restarted and same is tried again. I assume `error` should have the same semantics as throwing an exception.